components/ProductList/AllProductList.tsx

```diff
@@ -0,0 +1,78 @@
+import { ChangeEvent, useEffect, useState } from "react";
+import Link from "next/link";
+import Image from "next/image";
+import { ProductListTypes, ProductTypes } from "@/types/product";
+import searchIcon from "@/assets/images/ic_search.png";
+import ProductListItem from "./ProductListItem";
+
+function AllProductList({ products }: ProductListTypes) {
+  const [showedProducts, setShowedProducts] = useState<ProductTypes[]>([]);
+  const [order, setOrder] = useState<string>("createdAt");
+
+  useEffect(() => {
+    const sortedProducts = [...products].sort((a, b) => {
+      if (order === "createdAt") {
+        return (
+          new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
+        );
+      }
+      return (b as any)[order] - (a as any)[order];
+    });
```

Comment:
```suggestion
type SortOption  = "createdAt" | "favoriteCount";

const sortProducts = (products: ProductTypes[], order: SortOption) => {
  return [...products].sort((a, b) => {
    if (order === "createdAt") {
      return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
    }
    return b[order as keyof ProductTypes] - a[order as keyof ProductTypes];
  });
};

useEffect(() => {
  const sortedProducts = sortProducts(products, order);
  setShowedProducts(sortedProducts.slice(0, 10));
}, [order, products]);
```

* useEffect 안에 있던 식을 함수로 추출하고, 추출한 함수를 useEffect에서 호출만 하는 방법도 될것같아요